openwisp_monitoring/db/backends/elasticsearch/index.py

```diff
@@ -0,0 +1,109 @@
+import uuid
+
+from django.conf import settings
+from elasticsearch.exceptions import NotFoundError
+from elasticsearch.helpers import bulk
+from elasticsearch_dsl import Date, Document, InnerDoc, Nested, Q, Search
+from swapper import load_model
+
+# TODO: Remove this. Current Structure
+# Index Templates <-- Indexes <-- Documents <-- Points
+
+
+class Point(InnerDoc):
+    time = Date(required=True, default_timezone=settings.TIME_ZONE)
+    fields = Nested(dynamic=True, required=True, multi=True)
+
+
+class MetricIndex(Document):
+    tags = Nested(dynamic=True, required=False, multi=True)
+    points = Nested(Point)
+
+    class Index:
+        name = 'metric'
+        settings = {
+            'number_of_shards': 1,
+            'number_of_replicas': 0,
+            'lifecycle.name': 'default',
+            'lifecycle.rollover_alias': 'metric',
+        }
+
+
+def find_metric(client, index, tags, retention_policy=None, add=False):
+    search = Search(using=client, index=index)
+    if tags:
+        tags_dict = dict()
+        for key, value in tags.items():
+            tags_dict[f'tags.{key}'] = value
+        q = Q(
+            'nested',
+            path='tags',
+            query=Q(
+                'bool', must=[Q('match', **{k: str(v)}) for k, v in tags_dict.items()]
+            ),
+        )
+    else:
+        q = Q()
+    try:
+        result = list(search.query(q).execute())[0].meta
+        return result['id'], result['index']
+    except (NotFoundError, AttributeError, IndexError):
+        if add:
+            obj = add_doc(client, index, tags, retention_policy=retention_policy)
+            return obj['_id'], obj['_index']
+        return None
+
+
+def add_doc(client, key, tags, _id=None, retention_policy=None):
```

Comment:
`add_doc`?

Reply:
Yes. Suppose the `find_metric` function receives `device_data` metric with some tags. Then `device_data` would be an index whereas for every unique set of tags there shall be a new document created in the `device_data` index.

This function will check if index exists then add the document to it, else create the index and then add the document to it.

Let me know if I am able to explain it clearly :sweat_smile:

Reply:
it's the name of the function that troubles me, why is it named `add_doc`?

Reply:
Since the primary task of the function is to **add a document to elasticsearch**. I am not sure if the name is creating confusion since I am unable to think how :sweat_smile:. I can add a docstring to avoid any confusions.